### PR TITLE
Limit unauthorized file locking attempts

### PR DESCRIPTION
**Pull request by emgarten**
If NuGet is run as admin and crashes with a file lock open the file is left around. When a non-elevated process runs it is unable to access the file and will try forever. To make this friendlier to users file locking will give up after around 30 seconds and display a friendly error message describing how to unblock the situation.

//cc @yishaigalatzer @joelverhagen @zhili1208 @pspill

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs [hunk 1]
```diff
@@ -22,6 +23,9 @@ public static class ConcurrencyUtilities
                 throw new ArgumentNullException(nameof(filePath));
             }
 
+            // limit the number of unauthorized, this should be around 30 seconds.
+            var unauthorizedAttemptsLeft = 3000;
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs:27]: Can't we retry for a specific duration (using a stopwatch) rather than on a count?

Comment (emgarten) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs:27]: It's kind of odd with how System.IO works, but this exception is expected if another process has the lock and cleans up the file at the same time anther process is trying to get it. That means that the lock could be held for an hour (IOException) and then at the end it hits the access exception, in which case going by time would make it throw incorrectly.

**Comment by emgarten** [comment]
https://github.com/NuGet/Home/issues/2258

File: src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs [hunk 2]
```diff
@@ -48,7 +52,23 @@ public static class ConcurrencyUtilities
                     }
                     catch (UnauthorizedAccessException)
                     {
+                        token.ThrowIfCancellationRequested();
+
+                        if (unauthorizedAttemptsLeft < 1)
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs:57]: I love how we get this accurately, even though the number is 3000 :)

**Comment by yishaigalatzer** [comment]
:shipit: also need to let Jacques know we have a new resources dll to localize here so it gets picked up by the loc team

**Comment by yishaigalatzer** [comment]
and see if we need authoring changes in the vsix/msi

**Pull request closed without merging**